### PR TITLE
[feature] : quill add magnifier

### PR DESCRIPTION
**Pull request by demoYang**
## Description

*Add magnifier.* someone help me ?, i donot know how to pass all issue 

## Related Issues

- *Related #1504*

## Checklist

- [x] I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] I titled the PR using [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0).
- [x] I did not modify the `CHANGELOG.md` nor the package version in `pubspec.yaml` files.
- [x] All existing and new tests are passing.
- [x] I have run the commands in `./scripts/before_push.sh` and it all passed successfully

## Breaking Change

Does your PR require developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate that with a `!` in the title as explained in [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0)).
- [x] No, this is *not* a breaking change.

### CODE REVIEW

File: lib/src/widgets/others/text_selection.dart [hunk 1]
```diff
@@ -341,11 +352,12 @@ class EditorTextSelectionOverlay {
   /// Final cleanup.
   void dispose() {
     hide();
+    _magnifierInfo.dispose();
   }
 
   /// Builds the handles by inserting them into the [context]'s overlay.
   void showHandles() {
-    assert(_handles == null);
```
Comment (EchoEllet) [inline lib/src/widgets/others/text_selection.dart:348]: I'm not sure about this change. It might be there for a reason. `assert` will be only invoked in development and removed in production.

**Comment by singerdmx** [comment]
❌ /home/runner/work/flutter-quill/flutter-quill/test/bug_fix_test.dart: 1742 - Disable context menu after selection for desktop platform 1742 - Disable context menu after selection for desktop platform PointerDeviceKind.touch (failed)
  ══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════
  The following StateError was thrown while handling a gesture:
  Bad state: No element
  
  When the exception was thrown, this was the stack:
  #0      _Array.first (dart:core-patch/array.dart:51:5)
  #1      RenderEditableTextLine.getLineBoundary (package:flutter_quill/src/widgets/quill/text_line.dart:912:28)
  #2      RenderEditor.getLineAtOffset (package:flutter_quill/src/widgets/editor/editor.dart:1372:34)
  #3      EditorTextSelectionOverlay._buildMagnifier (package:flutter_quill/src/widgets/others/text_selection.dart:471:41)
  #4      EditorTextSelectionOverlay.showMagnifier (package:flutter_quill/src/widgets/others/text_selection.dart:403:7)
  #5      QuillRawEditorState.showMagnifier (package:flutter_quill/src/widgets/raw_editor/raw_editor_state.dart:1801:24)
  #6      _QuillEditorSelectionGestureDetectorBuilder.onSingleLongTapStart (package:flutter_quill/src/widgets/editor/editor.dart:567:13)
  #7      _EditorTextSelectionGestureDetectorState._handleLongPressStart (package:flutter_quill/src/widgets/others/text_selection.dart:[103](https://github.com/singerdmx/flutter-quill/actions/runs/9960808350/job/27522094543?pr=2026#step:14:105)3:36)

**Comment by CatHood0** [comment]
This is the part of the code that is throwing the error 

https://github.com/singerdmx/flutter-quill/blob/0dbe4c46853a41f8847ea272b20761430440f117/lib/src/widgets/quill/text_line.dart#L912

**Comment by demoYang** [comment]
and have a question : why not use  [TextSelectionGestureDetector](https://api.flutter.dev/flutter/widgets/TextSelectionGestureDetector-class.html) replace EditorTextSelectionGestureDetector in quill, 
TextSelectionGestureDetector have TapAndHorizontalDragGestureRecognizer that response tap and drag, so that  cursor can move when darg start and magnifier

**Comment by singerdmx** [comment]
> And i have a question : why not use [TextSelectionGestureDetector](https://api.flutter.dev/flutter/widgets/TextSelectionGestureDetector-class.html) replace EditorTextSelectionGestureDetector in quill, TextSelectionGestureDetector have TapAndHorizontalDragGestureRecognizer that response tap and drag, so that cursor and magnifier can move when darg start

EditorTextSelectionGestureDetector is probably in the old days. 
You can give it a try with TextSelectionGestureDetector.

**Comment by EchoEllet** [comment]
Can we separate this functionality into its own file/class?

**Comment by stepushchik** [comment]
https://github.com/singerdmx/flutter-quill/issues/2109

**Comment by stepushchik** [comment]
```
_magnifierController.show(
  context: context,
  below: magnifierConfiguration.shouldDisplayHandlesInMagnifier
      ? null
      : _handles![0],
  builder: (_) => builtMagnifier,
);
```

`Null check operator used on a null value in _showMagnifier(text_selection.dart:436)`: https://github.com/singerdmx/flutter-quill/issues/2109

**Comment by CatHood0** [comment]
@stepushchik i'll solving this. The PR will be submitted as soon as possible

File: lib/src/widgets/others/text_selection.dart (hunk 1 above)
Comment (EchoEllet) [reply lib/src/widgets/others/text_selection.dart:348]: @demoYang Can you explain why this has been removed? I do plan on reverting the change, it shouldn't caused any issues at least for the release mode.

File: lib/src/widgets/raw_editor/raw_editor_state.dart [hunk 2]
```diff
@@ -1340,33 +1347,36 @@ class QuillRawEditorState extends EditorState
 
   void _updateOrDisposeSelectionOverlayIfNeeded() {
     if (_selectionOverlay != null) {
-      if (!_hasFocus || textEditingValue.selection.isCollapsed) {
-        _selectionOverlay!.dispose();
-        _selectionOverlay = null;
-      } else {
```
Comment (EchoEllet) [inline lib/src/widgets/raw_editor/raw_editor_state.dart:1346]: Can you explain this change?

- Why `_hasFocus` has neglected?
- `textEditingValue.selection.isCollapsed` removed from the check?
- The `_selectionOverlay.dispose()` and `_selectionOverlay = null;` were removed?

File: lib/src/widgets/raw_editor/raw_editor_state.dart [hunk 3]
```diff
@@ -884,7 +884,13 @@ class QuillRawEditorState extends EditorState
     final oldSelection = controller.selection;
     controller.updateSelection(selection, ChangeSource.local);
 
+    if (_selectionOverlay == null) {
+      _selectionOverlay = _createSelectionOverlay();
+    } else {
+      _selectionOverlay!.update(textEditingValue);
+    }
     _selectionOverlay?.handlesVisible = _shouldShowSelectionHandles();
+    _selectionOverlay?.showHandles();
```
Comment (EchoEllet) [inline lib/src/widgets/raw_editor/raw_editor_state.dart:893]: Same for this change, if it's needed for this feature then shouldn't be hardcoded directly unless it's not related.

File: lib/src/widgets/raw_editor/raw_editor.dart [hunk 4]
```diff
@@ -84,4 +86,10 @@ abstract class EditorState extends State<QuillRawEditor>
   bool showToolbar();
 
   void requestKeyboard();
+
+  void showMagnifier(Offset positionToShow);
+
+  void updateMagnifier(Offset positionToShow);
+
+  void hideMagnifier();
```
Comment (EchoEllet) [inline lib/src/widgets/raw_editor/raw_editor.dart:94]: We will move those into their own abstract class

File: lib/src/widgets/quill/text_line.dart [hunk 5]
```diff
@@ -839,6 +839,10 @@ class RenderEditableTextLine extends RenderEditableBox {
         _getBoxes(TextSelection(baseOffset: 0, extentOffset: line.length - 1))
             .where((element) => element.top < lineDy && element.bottom > lineDy)
             .toList(growable: false);
+    if (lineBoxes.isEmpty) {
+      // Empty line, line box is empty
+      return TextRange.collapsed(position.offset);
+    }
```
Comment (EchoEllet) [inline lib/src/widgets/quill/text_line.dart:845]: Why this change was introduced? Does it solve a problem that is introduced with this feature?

File: lib/src/widgets/others/text_selection.dart [hunk 6]
```diff
@@ -366,8 +378,123 @@ class EditorTextSelectionOverlay {
...
+      ),
+    );
+  }
+
+  void _showMagnifier(MagnifierInfo initialMagnifierInfo) {
+    // 隐藏toolbar
+    if (toolbar != null) {
+      hideToolbar();
+    }
+    // 更新 magnifierInfo
```
Comment (EchoEllet) [inline lib/src/widgets/others/text_selection.dart:416]: English is the main supported language in this repository.

File: lib/src/widgets/others/text_selection.dart [hunk 7]
```diff
@@ -366,8 +378,123 @@ class EditorTextSelectionOverlay {
...
+  void showMagnifier(
+      TextPosition position, Offset offset, RenderEditor editor) {
+    _showMagnifier(
+      _buildMagnifier(
+        currentTextPosition: position,
+        globalGesturePosition: offset,
+        renderEditable: editor,
+      ),
+    );
+  }
```
Comment (EchoEllet) [inline lib/src/widgets/others/text_selection.dart:409]: The platform check needs to be more specific, it's not clear if it should check the Target platform regardless if the app is running on web browser or native app.

File: lib/src/widgets/others/text_selection.dart [hunk 8]
```diff
@@ -366,8 +378,123 @@ class EditorTextSelectionOverlay {
...
+    _magnifierController.show(
+      context: context,
+      below: magnifierConfiguration.shouldDisplayHandlesInMagnifier
+          ? null
+          : _handles![0],
+      builder: (_) => builtMagnifier,
+    );
+  }
+
+  void updateMagnifier(
```
Comment (EchoEllet) [inline lib/src/widgets/others/text_selection.dart:436]: Can you explain the difference between `updateMagnifier` and `_updateMagnifier`? Is the private one is for internal use? More details is needed

File: lib/src/widgets/others/text_selection.dart (hunk 1 above)
Comment (EchoEllet) [reply lib/src/widgets/others/text_selection.dart:348]: Reverted in #2338

Comment (EchoEllet) [reply lib/src/widgets/others/text_selection.dart:348]: The assert was removed to workaround the assertion error, which is something we will avoid in future PRs.

File: lib/src/widgets/quill/text_line.dart (hunk 5 above)
Comment (EchoEllet) [reply lib/src/widgets/quill/text_line.dart:845]: @demoYang This issue is not solved.
